### ISSUE TITLE
How to retrieve AuditEntry records for parent and associated children?

### ISSUE THREAD

**Issue by mhsimkin** [zzzprojects/EntityFramework-Plus#151]
Hi:

I'm trying to figure out the best way to solve an issue.

I have extended the ASP.Net Identity to allow for auditing.  All that is  working great.

Below is an excerpted version of the ASP.Net IdentityUser class:

    public class IdentityUser<TKey, TLogin, TRole, TClaim> : IUser<TKey>
        where TLogin : IdentityUserLogin<TKey>
        where TRole : IdentityUserRole<TKey>
        where TClaim : IdentityUserClaim<TKey>
    {
        public virtual ICollection<TRole> Roles { get; private set; }
        public virtual ICollection<TClaim> Claims { get; private set; }
        public virtual ICollection<TLogin> Logins { get; private set; }
		public virtual TKey Id { get; set; }
    }

When I retrieve the audit entry for an IdentityUser change, I would also like to retrieve all the audit entries for any of the associated collections that might of changed at the same time.

I'm currently using the below statement to retrieve the audit entries for a user

    DbContext.AuditEntries.Where(user)

How can I change this to include all the audit entry records for the associated collections?  Can this be done without extending the AuditEntry class?

I have reviewed the documentation on [Audit Customization](https://github.com/zzzprojects/EntityFramework-Plus/wiki/EF-Audit-%7C-Entity-Framework-Audit-Trail-Context-and-Track-Changes#audit-customization) and I'm thinking that I need to create a Custom AuditEntry class, that has a new string property called SourceId.

Then in the overridden DbContext::Save method, I should hook of the AuditEntryFactory to create a new instance of my custom class and set the SourceId property.

Does this all seem like it will work?

Thanks

-marc

**Comment by mhsimkin** [comment]
I'm attempting to get this to work using the Audit Customization approach.  I've run into a problem.

I created a new class

`    public class PdsAuditEntry : AuditEntry
    {
        [Required]
        [MaxLength(128)]
        public string SourceID { get; set; }
    }
`
I updated the DbSet in my context to be my new class.  I added code to the onModelBuilding to change the table name to be PdsAuditEntries table.

`modelBuilder.Entity<PdsAuditEntry>().ToTable("PdsAuditEntries");`

Inside my overridden DbContext::Save method I have this code:

`            var audit = new Z.EntityFramework.Plus.Audit();
            audit.Configuration.IgnorePropertyUnchanged = true;
            audit.Configuration.AuditEntryFactory = args => new PdsAuditEntry() { SourceID = GetSourceId(dbContext) };
            audit.PreSaveChanges(dbContext);
`

When I attempt to save records to the database, EF is generating this SQL

`
exec sp_executesql N'INSERT [dbo].[AuditEntries]([EntitySetName], [EntityTypeName], [State], [StateName], [CreatedBy], [CreatedDate])
VALUES (@0, @1, @2, @3, @4, @5)
SELECT [AuditEntryID]
FROM [dbo].[AuditEntries]
WHERE @@ROWCOUNT > 0 AND [AuditEntryID] = scope_identity()',N'@0 nvarchar(255),@1 nvarchar(255),@2 int,@3 nvarchar(255),@4 nvarchar(255),@5 datetime2(7)',@0=N'IdentityUserRoles',@1=N'IdentityUserRole',@2=0,@3=N'EntityAdded',@4=N'BNHQ\spmys',@5='2017-04-07 16:20:24.4201129'
`
What jumps out at me, besides using the wrong table name, it is missing mu new column, SourceID.

What did I miss in the code?  I know that the AudityEntryProperty still refers to the AuditEntry class.

I didn't set up the global AuditManager.DefaultConfiguration.AuditEntryFactory since I was doing this per instance, as I need to walk the through the entities being changed to find the parent entity, which contains the Id I need to use for Source Id.

Do I need to set up the global handler?  Does the global handler get passed the current DbContext?

Any pointers on where I should look.  Any idea what I missed?

Thanks

marc

**Comment by JonathanMagnan** [comment]
Hello @mhsimkin ,

I believe the inheritance is not set correctly.

I recommend you modifying `DbSet<AuditEntry>` to `DbSet<PdsAuditEntry>`

I believe this should fix your issue and you will find the value in the PdsAuditEntry table.

Best Regards,

Jonathan

**Comment by mhsimkin** [comment]
I originally did that.  I created an interface called IAudit, which is defined as:
```
    public interface IAudit
    {
        DbSet<PdsAuditEntry> AuditEntries { get; set; }
        DbSet<AuditEntryProperty> AuditEntryProperties { get; set; }
    }
```
I then added that interface to my classes that are derived from DbContext, for example:

```
public class ClaimsDbContext : IdentityDbContext<User, Role, string, IdentityUserLogin, IdentityUserRole, IdentityUserClaim>, IAudit
{
        public DbSet<PdsAuditEntry> AuditEntries { get; set; }
        public DbSet<AuditEntryProperty> AuditEntryProperties { get; set; }
}
```
I even tried creating an AuditEntryProperty derived class.  I updated my interface and all my DbContext classes to use my AuditEntryProperty derived class.  

I still get an error about an invalid object name 'dbo.AuditEntries' when I attempt to save any of the audit info to the database.

I think the issue is being caused by both the AuditEntry and AuditEntryProperty classes referring to each other.  For example:
```
public class AuditEntryProperty
{
     public int AuditEntryPropertyID { get; set; }
     public int AuditEntryID { get; set;}
     public AuditEntry Parent { get; set; }
}
```
I think you will need to do something similar to how ASP.NET Identity structures their classes to allow for customization:
```
public class AuditEntryProperty : AuditEntryProperty<AuditEntry>
{
}

public class AuditEntryProperty<TEntry>
{
     public int AuditEntryPropertyID { get; set; }
     public int AuditEntryID { get; set;}
     public TEntry Parent { get; set; }
}

public class AuditEntry : AuditEntry<AuditEntryProperty>
{
}

public class AuditEntry<TProperty>
{
     public int AuditEntryID { get; set; }
     public List<TProperty> Properties { get; set; }
}
```
I would attempt this myself to see if it works, however, after cloning the code repository and restoring the Nuget packages.  I can't get the solution to build.  I'm getting all sorts of errors from missing namespaces to .NET Core missing.

I'm willing to help with some code, provided I can get everything to build.  I'm also willing to provide a sample implementation that covers changeset as well as using a GUID as the key.

Thanks

-marc

**Comment by mhsimkin** [comment]
I tried to get something to work by defining interfaces for IAuditEntry and IAuditEntryProperty.  I seem to be headed down a rabbit hole.  I'm just going to give up on this as I don't have time over the next few days.

Ultimately what I need to be able to do are:

1) For a given parent record, say ASP.NET Identity IdentityUser, find all changes to the IdentityUser record and all it's children (IdentityUserRole, IdentityUserLogin, IdentityUserClaim).

2) For a specific change find all the related audit records.  For example, if I add a user to a role the IdentityUser record is modified and a new IdentityUserRole record is added.  I determined by doing some SQL queries, that these are treated as two separate audit entries.  

When I was trying to extend the AuditEntry class, it was so that I can resolve items 1 & 2.  If there is a way I can extend the classes and resolve item #2, that would be enough.

Thanks

marc

**Comment by JonathanMagnan** [comment]
Hello @mhsimkin ,

> I still get an error about an invalid object name 'dbo.AuditEntries' when I attempt to save any of the audit info to the database.

I believe this is because the database has not been re-created with new changes.

Here is a working example for customizing Audit

```csharp
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.Data.Entity;
using System.Windows.Forms;
using Z.EntityFramework.Plus;

namespace Z.Request.CustomAudit
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();

            // CLEAR
            using (var ctx = new CurrentContext())
            {
                ctx.AuditEntries.RemoveRange(ctx.AuditEntries);
                ctx.AuditEntryProperties.RemoveRange(ctx.AuditEntryProperties);
                ctx.Lefts.RemoveRange(ctx.Lefts);
                ctx.Rights.RemoveRange(ctx.Rights);
            }

            // SEED

            // TEST
            AuditManager.DefaultConfiguration.AutoSavePreAction = (context, audit) =>
                (context as CurrentContext).AuditEntries.AddRange(audit.Entries);

            using (var dbContext = new CurrentContext())
            {
                var audit = new Audit();
                audit.Configuration.IgnorePropertyUnchanged = true;
                audit.Configuration.AuditEntryFactory = args => new PdsAuditEntry {SourceID = GetSourceId(dbContext)};

                var left = dbContext.Lefts.Add(new Left {IntColumn = 1});

                dbContext.SaveChanges(audit);
            }
        }

        public string GetSourceId(DbContext context)
        {
            return Guid.NewGuid().ToString();
        }

        public class CurrentContext : DbContext
        {
            public CurrentContext() : base("CodeFirstEntities")
            {
            }

            public DbSet<AuditEntry> AuditEntries { get; set; }
            public DbSet<AuditEntryProperty> AuditEntryProperties { get; set; }
            public DbSet<Left> Lefts { get; set; }
            public DbSet<Right> Rights { get; set; }

            protected override void OnModelCreating(DbModelBuilder modelBuilder)
            {
                modelBuilder.Entity<AuditEntry>().ToTable("PdsAuditEntries");
                modelBuilder.Entity<AuditEntryProperty>().ToTable("PdsAuditEntryProperties");

                base.OnModelCreating(modelBuilder);
            }
        }

        public class PdsAuditEntry : AuditEntry
        {
            [Required]
            [MaxLength(128)]
            public string SourceID { get; set; }
        }

        public class Left
        {
            public int ID { get; set; }
            public int IntColumn { get; set; }
            public List<Right> Rights { get; set; }
        }

        public class Right
        {
            public int ID { get; set; }
            public int IntColumn { get; set; }
            public Left Left { get; set; }
        }
    }
}
```

**Comment by mhsimkin** [comment]
Jonathan, thank you.  I will give this a try tomorrow.  I need to figure out what is going wrong.

**Comment by mhsimkin** [comment]
Hi Jonathan:

Thank you for all your help.  I've been able to get this working.  Your example helped me find and resolve the issues I was seeing.

To summarize, in case anyone else is trying to do the same thing.

1) I had migrations turned off.  It seems that EF is smart enough to place derived types and base types in the same table. It is adding a new column called Discriminator, so that it can track the actual class used to create the entry.

2) There was a mismatch in the capitalization of the property name in my C# code vs the table in the database.

3) I was trying to use my new derived (PdsAuditEntry) class in the DbSet when I should of been using the base (AuditEntry) class.

I ended up working out a way to create a change set id so that I can retrieve all the entries that changed at the same time.

My AuditEntry derived class looks like this:

    public class PdsAuditEntry : AuditEntry
    {
        public PdsAuditEntry(Guid changeSetId) : base()
        {
            ChangeSetID = changeSetId;
        }

        [Required]
        public Guid ChangeSetID { get; set; }
    }

Right before the call to create a new instance of the Audit object I assign a new Guid to a variable, and then pass that variable to the AuditEntryFactory call:

            // generate the change set id for this change.
            var changeSetId = Guid.NewGuid();

            var audit = new Audit();
            audit.Configuration.IgnorePropertyUnchanged = true;
            audit.Configuration.AuditEntryFactory = args => new PdsAuditEntry(changeSetId);
            audit.PreSaveChanges(dbContext);

Once again, thank you for your help and for teaching something new about EF that I didn't know.

-marc
